src/main/java/ch/njol/skript/Skript.java

```diff
@@ -679,22 +675,63 @@ protected void afterErrors() {
+									// Test that requires package access.
+									classes.add(Class.forName("ch.njol.skript.variables.FlatFileStorageTest"));
+									size = classes.size();
+									for (Class<?> clazz : classes) {
+										String test = clazz.getName();
+										SkriptJUnitTest.lastJUnitTest = test;
+										info("Running JUnit test '" + test + "'");
+										Result junit = JUnitCore.runClasses(clazz);
+										TestTracker.testStarted("JUnit: '" + test + "'");
+										shutdown += SkriptJUnitTest.getDelay();
+										SkriptJUnitTest.setDelay(0);
+										tests += junit.getRunCount();
+										milliseconds += junit.getRunTime();
+										ignored += junit.getIgnoreCount();
+										if (junit.getFailureCount() > 0) {
+											fails += junit.getFailureCount();
+											junit.getFailures().forEach(failure -> {
+												String message = failure.getMessage() == null ? "" : " " + failure.getMessage();
+												TestTracker.testFailed("'" + test + "': " + message);
+												Skript.exception(failure.getException(), "JUnit test '" + failure.getTestHeader() + " failed.");
+											});
+										}
+										SkriptJUnitTest.lastJUnitTest = null;
+									}
+								} catch (IOException e) {
+									Skript.exception(e, "Failed to execute JUnit runtime tests.");
+								} catch (ClassNotFoundException e) {
+									// Should be the Skript test jar gradle task.
+									assert false;
+								}
+								if (ignored > 0)
+									Skript.warning("There were " + ignored + " ignored test cases! This can mean they are not properly setup in order in that class!");
+								
+								info("Completed " + tests + " JUnit tests in " + size + " classes with " + fails + " failures in " + milliseconds + " milliseconds.");
 							}
-							info("Testing done, shutting down the server.");
-							// Delay server shutdown to stop the server from crashing because the current tick takes a long time due to all the tests
-							Bukkit.getScheduler().runTaskLater(Skript.this, () -> {
-								Bukkit.getServer().shutdown();
-							}, 5);
-
 						}
-						return;
-					}
-				}, 100);
+						info("Collecting results to " + TestMode.RESULTS_FILE);
+						String results = new Gson().toJson(TestTracker.collectResults());
+						try {
+							Files.write(TestMode.RESULTS_FILE, results.getBytes(StandardCharsets.UTF_8));
+						} catch (IOException e) {
+							Skript.exception(e, "Failed to write test results.");
+						}
+						long seconds = TimeUnit.MILLISECONDS.toSeconds(shutdown);
+						info("Testing done, shutting down the server in " + seconds + " second " + (seconds == 1 ? "" : "s") + "...");
+						// Delay server shutdown to stop the server from crashing because the current tick takes a long time due to all the tests
+						Bukkit.getScheduler().runTaskLater(Skript.this, () -> {
+							Bukkit.getServer().shutdown();
+						}, seconds * 20);
```

Comment:
Does all the testing have to take the place on the same tick? If not, it could be limited to X tests per tick.

Reply:
As mentioned in comment https://github.com/SkriptLang/Skript/pull/4979#discussion_r1059488729